services/fileuploader/fileuploader.go

```diff
@@ -0,0 +1,160 @@
+// Useful for tests.
+func NewStaticProvider(storageSettings map[string]StorageSettings) Provider {
+	s := &provider{
+		init:            make(chan struct{}),
+		storageSettings: storageSettings,
+	}
+	close(s.init)
+	return s
+}
+
+// NewDefaultProvider creates a new provider that operates against the default storage settings populated from the env.
+// Useful for tests that populate settings from env.
+func NewDefaultProvider() Provider {
+	d := &defaultProvider{}
+	return d
+}
+
+type provider struct {
+	onceInit        sync.Once
+	init            chan struct{}
+	storageSettings map[string]StorageSettings
+}
+
+func (p *provider) GetFileManager(workspaceID string) (filemanager.FileManager, error) {
+	<-p.init
+	settings, ok := p.storageSettings[workspaceID]
+	if !ok {
+		return nil, fmt.Errorf("no storage settings found for workspace: %s", workspaceID)
+	}
+	return filemanager.DefaultFileManagerFactory.New(&filemanager.SettingsT{
+		Provider: settings.Bucket.Type,
+		Config:   settings.Bucket.Config,
+	})
+}
+
+func (p *provider) GetStoragePreferences(workspaceID string) (backendconfig.StoragePreferences, error) {
+	<-p.init
+	var prefs backendconfig.StoragePreferences
+	settings, ok := p.storageSettings[workspaceID]
+	if !ok {
+		return prefs, fmt.Errorf("no storage settings found for workspace: %s", workspaceID)
+	}
+	return settings.Preferences, nil
+}
+
+// updateLoop uses backend config to retrieve & keep up-to-date the storage settings of all workspaces.
+func (p *provider) updateLoop(ctx context.Context, backendConfig backendconfig.BackendConfig) {
+	ch := backendConfig.Subscribe(ctx, backendconfig.TopicBackendConfig)
+
+	settings := make(map[string]StorageSettings)
+
+	var bucket backendconfig.StorageBucket
+	var preferences backendconfig.StoragePreferences
+
+	for ev := range ch {
+		configs := ev.Data.(map[string]backendconfig.ConfigT)
+		for _, c := range configs {
+			if c.Settings.DataRetention.UseSelfStorage {
+				settings := c.Settings.DataRetention.StorageBucket
+				defaultBucket := getDefaultBucket(ctx, settings.Type)
+				bucket = overrideWithSettings(defaultBucket.Config, settings, c.WorkspaceID)
+			} else {
+				bucket = getDefaultBucket(ctx, config.GetString("JOBS_BACKUP_STORAGE_PROVIDER", "S3"))
+			}
+			preferences = c.Settings.DataRetention.StoragePreferences
+			settings[c.WorkspaceID] = StorageSettings{
+				Bucket:      bucket,
+				Preferences: preferences,
+			}
+		}
+		p.onceInit.Do(func() {
+			close(p.init)
+		})
+		p.storageSettings = settings
+	}
+
+	p.onceInit.Do(func() {
+		close(p.init)
+	})
+}
+
+type defaultProvider struct{}
+
+func (*defaultProvider) GetFileManager(_ string) (filemanager.FileManager, error) {
+	defaultConfig := getDefaultBucket(context.Background(), config.GetString("JOBS_BACKUP_STORAGE_PROVIDER", "S3"))
+	return filemanager.DefaultFileManagerFactory.New(&filemanager.SettingsT{
+		Provider: defaultConfig.Type,
+		Config:   defaultConfig.Config,
+	})
```

Comment:
please add a test scenario for this function

Reply:
Done